### DISCUSSION TITLE
What's the difference between PDM and Virtual Envs?

### DISCUSSION THREAD

**Discussion by robml** [pdm-project/pdm#894] [General]
## Is your feature request related to a problem? Please describe.

So I understand from using PDM it doesn't require to be "activated" in a virtual environment to use, I simply go to the project folder and can start using everything there with all my libraries also there. My question is, what is the difference from instantiating a "venv" in your project folder and holding everything there? I am still trying to understand how package managers work here. Also does PDM create a physical copy of each library or does it point to the existing library if it is already installed on the system? Would it not be more redundant to have the libraries stored per project if I can reuse a common set of libraries in a single virtual env that I can use across projects? These are just a few of my questions as I find myself trying to find a set up that works for myself and our team.

## Describe the solution you'd like

Just trying to understand.

**Comment by pawamoy** [comment]
Functionally speaking (can wa say that in english?), using PDM or Poetry is not that different than explicitely using a virtualenv. Poetry hides it while PDM just does not use virtualenvs. The advantage I see in PDM is that you don't have to activate and deactivate your venv anymore (you just run `pdm run ...`), and you don't even *have* a virtualenv. It's simpler. Packages are available in the `__pypackages__` folder and that's it, Python can find them.

> Also does PDM create a physical copy of each library or does it point to the existing library if it is already installed on the system?

PDM has the `install_cache` feature which keeps installed packages in a cache folder, and create symlinks (or .pth files) to them in your `__pypackages__` folder. It makes the installation faster, and saves tremendous amounts of space on your disk :slightly_smiling_face:

**Reply by robml** [reply]
This is actually very intuitive thank you, I had no idea it worked in such a way. Is this cache folder somewhere else on the system I assume then? That way my symlinks can all point to the same place? All in all the above definitely seems to take care a lot for you then, and has made me appreciate PDM more.

**Reply by pawamoy** [reply]
On Linux the packages cache folder is located at `~/.cache/pdm/packages`. PDM takes care of creating the symlinks when running `pdm install`. It's indeed one of the features that makes PDM so much more enjoyable than other alternatives ^^

**Reply by frostming** [reply]
I'd like to refer to several articles:
- https://frostming.com/2021/01-22/introducing-pdm/#the-problems-with-virtual-environments
- https://laike9m.com/blog/best-python-development-setup-for-2022-and-beyond,144/#pdm

**Comment by pawamoy** [comment]
@frostming I'll go ahead and move this to a discussion, let me know if you'd prefer I don't do that.

**Reply by robml** [reply]
Thank you for doing so, I had no idea the Discussions tab existed here. Seems more appropriate.
